src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -539,6 +539,67 @@ void ParallelScavengeHeap::object_iterate(ObjectClosure* cl) {
   old_gen()->object_iterate(cl);
 }
 
+// The HeapBlockClaimer is used during parallel iteration over the heap,
+// allowing workers to claim heap blocks, gaining exclusive rights to these blocks.
+// The eden and survivor spaces are treated as single blocks as it is hard to divide
+// these spaces.
+// The old spaces are divided into serveral fixed-size blocks.
+class HeapBlockClaimer : public StackObj {
+  // Index of iterable block, negative values for indexes of young generation spaces,
+  // zero and positive values for indexes of blocks in old generation space.
+  ssize_t _claimed_index;
+ public:
+  static const ssize_t EdenIndex = -2;
+  static const ssize_t SurvivorIndex = -1;
+
+  HeapBlockClaimer() : _claimed_index(EdenIndex) { }
+  // Claim the block and get the block index.
+  bool claim_and_get_block(ssize_t* block_index) {
+    assert(block_index != NULL, "Invalid index pointer");
+    *block_index = Atomic::fetch_and_add(&_claimed_index, 1);
+    ssize_t iterable_blocks = (ssize_t)ParallelScavengeHeap::heap()->old_gen()->iterable_blocks();
+    if (*block_index >= iterable_blocks) {
+      return false;
+    }
+    return true;
+  }
+};
+
+void ParallelScavengeHeap::object_iterate_parallel(ObjectClosure* cl,
+                                                   HeapBlockClaimer* claimer) {
+  ssize_t block_index;
+  // Iterate until all blocks are claimed
+  while (claimer->claim_and_get_block(&block_index)) {
+    if (block_index == HeapBlockClaimer::EdenIndex) {
+      young_gen()->eden_space()->object_iterate(cl);
+    } else if (block_index == HeapBlockClaimer::SurvivorIndex) {
+      young_gen()->from_space()->object_iterate(cl);
+      young_gen()->to_space()->object_iterate(cl);
+    } else {
+      old_gen()->block_iterate(cl, (size_t)block_index);
+    }
+  }
```

Comment:
The structure seems to suggest that each `if` case could be hit in every iteration, but actually only the third case will be revisited more than once. I wonder if this could be made more explicit.

Reply:
Thanks Albert! Correct, the OldGen part will be revisit much for frequently.  do you have any clue to optimize this? I was considering the marco like "likely()" but I am not sure whether it is acceptable for hotspot.

Reply:
I think readability is the issue here; below is my attempt with the new `claim_and_get_block`. Don't think there's any performance diff though.

```        
size_t block_index = claimer->claim_and_get_block();
if (block_index == HeapBlockClaimer::EdenIndex) {
  young_gen()->eden_space()->object_iterate(cl);
  block_index = claimer->claim_and_get_block();
}
if (block_index == HeapBlockClaimer::SurvivorIndex) {
  young_gen()->from_space()->object_iterate(cl);
  young_gen()->to_space()->object_iterate(cl);
  block_index = claimer->claim_and_get_block();
}
while (block_index != HeapBlockClaimer::InvalidIndex) {
  old_gen()->object_iterate_block(cl, block_index - HeapBlockClaimer::NumNonOldGenClaims);
  block_index = claimer->claim_and_get_block();
}
```        
PS: I didn't test the code, so please take it as a quick sketch.

Reply:
Hi Albert, 
Thanks for suggestion and I get your point. I have a little concern with this change:  it presumes that EdenIndex and SurvivorIndex are fix to be valued as the begining of the indices (0 and 1), otherwise there are chances that youngGen may never get iterated. I am not sure whether this assumption should be kept (or at least add the info in comments?) , although the current implementation of using size_t for indices follows the assumption.

Thanks,
Lin

Reply:
> it presumes that EdenIndex and SurvivorIndex are fix to be valued as the begining of the indices (0 and 1)

I don't think that's problematic; the assumption, the iteration follows the order of eden, from, to, and old-gen, actually makes the code easier to read, I believe. Having said that, I have no strong opinion on this; it's up to you.

Reply:
Keeping the order of iteration following eden,from,to and old-gen sounds reasonable. Thanks for explaination, will made the change in the new commit.

Reply:
Thank you for the change.